src/uct/tcp/tcp_sockcm.c

```diff
@@ -46,76 +56,62 @@ static uct_cm_ops_t uct_tcp_sockcm_ops = {
     .ep_create        = uct_tcp_sockcm_ep_create
 };
 
-static void uct_tcp_close_ep(uct_tcp_sockcm_ep_t *ep)
+static void uct_tcp_sockcm_close_ep(uct_tcp_sockcm_ep_t *ep)
 {
     ucs_list_del(&ep->list);
     ucs_async_remove_handler(ep->fd, 1);
-    close(ep->fd);
-    ep->fd = -1;
+    ucs_close_fd(&ep->fd);
     UCS_CLASS_DELETE(uct_tcp_sockcm_ep_t, ep);
 }
 
+static inline void uct_tcp_sockcm_ep_handle_status(uct_tcp_sockcm_ep_t *ep,
+                                                   ucs_status_t status,
+                                                   int events, const char *reason)
+{
+    if (status != UCS_OK) {
+        ucs_trace("removing ep %p (fd=%d state=%d events=%d) async events handler "
+                  "because %s: %s ",
+                  ep, ep->fd, ep->state, events, reason, ucs_status_string(status));
+        ucs_async_modify_handler(ep->fd, 0);
+    }
+}
+
 void uct_tcp_sa_data_handler(int fd, int events, void *arg)
 {
     uct_tcp_sockcm_ep_t *ep = (uct_tcp_sockcm_ep_t*)arg;
     ucs_status_t status;
 
     ucs_assertv(ep->fd == fd, "ep->fd %d fd %d, ep_state %d", ep->fd, fd, ep->state);
+    ucs_assertv(ucs_popcount(events) == 1, "events=%d", events);
```

Comment:
why do we assume there will be only one event?

Reply:
@yosefe why would there be more? this function handles only one.
more than one event can arrive only from ucs_async_poll() which is called from ucs_async_check_miss() when the async mode is UCS_ASYNC_MODE_POLL, which isn't used in uct or ucp now.

Reply:
since events is bitmap, it can potentially have more than one event, otherwise it would have been a single value.

Reply:
@yosefe right but can it happen to tcp_sockcm in this function? in which flow can this happen?
do you mean that this function should be reworked to handle more than one event in a single call?

Reply:
according to ucs/async callback API, this function could be called with more than one event. And according to the code, it can happen if async/thread for example.
the function should check for each possible event bit separately.

Reply:
i will add a check but i dont see a flow in ucp/ucs which enables calling this function with more than 1 event set. can you please describe an existing flow in ucx that can cause this?

Reply:
ucs_event_set_wait() calls epoll_wait() which returns more than one event -> ucs_event_set_map_to_events() translates to multiple events -> ucs_async_thread_ev_handler() called with multiple events-> ucs_async_dispatch_handlers() -> ucs_async_handler_dispatch() -> ucs_async_handler_invoke() -> the callback